lib/puppet/util/http_proxy.rb

```diff
@@ -11,6 +20,36 @@ def self.http_proxy_env
     return nil
   end
 
+  def self.no_proxy?(dest)
+    unless no_proxy_env = ENV["no_proxy"] || ENV["NO_PROXY"]
+      return false
+    end
+
+    unless dest.is_a? URI
+      begin
+        dest = URI.parse(dest)
+      rescue URI::InvalidURIError
+        return false
+      end
+    end
+
+    no_proxy_env.split(/\s*,\s*/).each do |d|
+      host, port = d.split(':')
+      host = Regexp.escape(host).gsub('\*', '.*')
+
+      #If this no_proxy entry specifies a port, we want to match it against
+      #the destination port.  Otherwise just match hosts.
+      no_proxy_regex  = port ? %r(^#{host}:#{port}$) : %r(^#{host}$)
+      dest_string     = port ? "#{dest.host}:#{dest.port}" : "#{dest.host}"
```

Comment:
Thanks, much more readable to me! :+1:

Now that I grokked it, I come to believe that a

    if port
      no_proxy_regex = ...
      dest_string = ...
    else
      ...
    end

would be even clearer. Thoughts?